sync/syncer.go

```diff
@@ -145,97 +148,175 @@ func (s *Syncer) Synchronise() {
 				return
 			}
 		}
-		s.ValidateLayer(lyr)
+
+		mu.Lock()
+		go func() {
+			s.ValidateLayer(lyr) //run one at a time
+			mu.Unlock()
+		}()
 	}
 }
 
 func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.Layer, error) {
-	blockIds, err := s.getLayerBlockIDs(types.LayerID(currenSyncLayer))
+
+	//fetch layer hash from each peer
+	m, err := s.fetchLayerHashes(currenSyncLayer)
+
+	if err != nil {
+		s.Error("could not get LayerHashes for layer: %v", currenSyncLayer)
+		return nil, err
+	}
+
+	//fetch ids for each hash
+	blockIds, err := s.fetchLayerBlockIds(m, currenSyncLayer)
 	if err != nil {
 		s.Error("could not get layer block ids %v", currenSyncLayer, err)
 		return nil, err
 	}
-	output := FetchBlocks(s, blockIds)
+
+	blocksArr, err := s.fetchFullBlocks(blockIds)
+	if err != nil {
+		s.Error("could not get layer blocks %v", currenSyncLayer, err)
+		return nil, err
+	}
+
+	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
+}
+
+func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
+	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		mb := out.(*types.MiniBlock)
-		foundTxs, missing := s.GetTransactions(mb.TxIds)
-		fetchedTxs := s.fetchTxs(missing)
-		txs := make([]*types.SerializableTransaction, len(mb.TxIds))
-		for _, t := range mb.TxIds {
-			if tx, ok := foundTxs[t]; ok {
-				txs = append(txs, tx)
-			} else {
-				txs = append(txs, fetchedTxs[t])
-			}
+
+		//sync Transactions
+		txs, err := s.Txs(mb)
+		if err != nil {
+			s.Warning(fmt.Sprintf("failed fetching block %v transactions ", err))
```

Comment:
what happens when we fail to fetch some part of the block? is this ok to return only a partial list? how will the receiving end act if only a partial list is returned

Reply:
this is not handled yet, sync will just keep tring to fetch the missing information